pkg/server/application_api/application_api_test: TestClusterResetSQLStats failed

pkg/server/application_api/application_api_test.TestClusterResetSQLStats [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12451844?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12451844?buildTab=artifacts#/) on master @ [974b43dbdc6fb612bc2e1c35a56429d2b779102a](https://github.com/cockroachdb/cockroach/commits/974b43dbdc6fb612bc2e1c35a56429d2b779102a):


```
=== RUN   TestClusterResetSQLStats
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/03e3c3f86914ef5ebcca1a9ffbd26abe/logTestClusterResetSQLStats1808624681
    test_log_scope.go:81: use -show-logs to present logs inline
    stats_test.go:392: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/03e3c3f86914ef5ebcca1a9ffbd26abe/logTestClusterResetSQLStats1808624681
--- FAIL: TestClusterResetSQLStats (66.24s)
=== RUN   TestClusterResetSQLStats/flushed=true
    test_server_shim.go:152: automatically injected virtual cluster under test; see comment at top of test_server_shim.go for details.
    stats_test.go:358: expected to find stats for at least one statement, but found: 0
    --- FAIL: TestClusterResetSQLStats/flushed=true (39.07s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/unowned
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterResetSQLStats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-33058

pkg/server/application_api/application_api_test.TestClusterResetSQLStats [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12451845?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12451845?buildTab=artifacts#/) on master @ [974b43dbdc6fb612bc2e1c35a56429d2b779102a](https://github.com/cockroachdb/cockroach/commits/974b43dbdc6fb612bc2e1c35a56429d2b779102a):


```
=== RUN   TestClusterResetSQLStats
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/03e3c3f86914ef5ebcca1a9ffbd26abe/logTestClusterResetSQLStats367825252
    test_log_scope.go:81: use -show-logs to present logs inline
    stats_test.go:392: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/03e3c3f86914ef5ebcca1a9ffbd26abe/logTestClusterResetSQLStats367825252
--- FAIL: TestClusterResetSQLStats (91.10s)
=== RUN   TestClusterResetSQLStats/flushed=true
    test_server_shim.go:152: automatically injected virtual cluster under test; see comment at top of test_server_shim.go for details.
    stats_test.go:358: expected to find stats for at least one statement, but found: 0
    --- FAIL: TestClusterResetSQLStats/flushed=true (45.78s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterResetSQLStats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/server/application_api/application_api_test.TestClusterResetSQLStats [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12468637?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12468637?buildTab=artifacts#/) on master @ [cedd409f2115655efe2c529f6052f3056673b277](https://github.com/cockroachdb/cockroach/commits/cedd409f2115655efe2c529f6052f3056673b277):


```
=== RUN   TestClusterResetSQLStats
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/03e3c3f86914ef5ebcca1a9ffbd26abe/logTestClusterResetSQLStats4231910339
    test_log_scope.go:81: use -show-logs to present logs inline
    stats_test.go:392: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/03e3c3f86914ef5ebcca1a9ffbd26abe/logTestClusterResetSQLStats4231910339
--- FAIL: TestClusterResetSQLStats (63.57s)
=== RUN   TestClusterResetSQLStats/flushed=true
    test_server_shim.go:152: automatically injected virtual cluster under test; see comment at top of test_server_shim.go for details.
    stats_test.go:358: expected to find stats for at least one statement, but found: 0
    --- FAIL: TestClusterResetSQLStats/flushed=true (31.24s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterResetSQLStats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/server/application_api/application_api_test.TestClusterResetSQLStats [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12486040?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12486040?buildTab=artifacts#/) on master @ [9e439be58101013a694e0362f774082b6b844fee](https://github.com/cockroachdb/cockroach/commits/9e439be58101013a694e0362f774082b6b844fee):


```
=== RUN   TestClusterResetSQLStats
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/03e3c3f86914ef5ebcca1a9ffbd26abe/logTestClusterResetSQLStats662560716
    test_log_scope.go:81: use -show-logs to present logs inline
    stats_test.go:392: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/03e3c3f86914ef5ebcca1a9ffbd26abe/logTestClusterResetSQLStats662560716
--- FAIL: TestClusterResetSQLStats (68.87s)
=== RUN   TestClusterResetSQLStats/flushed=true
    test_server_shim.go:165: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    stats_test.go:358: expected to find stats for at least one statement, but found: 0
    --- FAIL: TestClusterResetSQLStats/flushed=true (40.30s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterResetSQLStats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/server/application_api/application_api_test.TestClusterResetSQLStats [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12497573?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12497573?buildTab=artifacts#/) on master @ [0496181c886c305f261598f90f0b226f9940fdac](https://github.com/cockroachdb/cockroach/commits/0496181c886c305f261598f90f0b226f9940fdac):


```
=== RUN   TestClusterResetSQLStats
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/44ea9a87ccb65e6e113f3631d444aaca/logTestClusterResetSQLStats856179153
    test_log_scope.go:81: use -show-logs to present logs inline
    stats_test.go:392: -- test log scope end --
--- FAIL: TestClusterResetSQLStats (20.17s)
=== RUN   TestClusterResetSQLStats/flushed=true
    test_server_shim.go:165: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    stats_test.go:358: expected to find stats for at least one statement, but found: 0
    --- FAIL: TestClusterResetSQLStats/flushed=true (10.75s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterResetSQLStats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/server/application_api/application_api_test.TestClusterResetSQLStats [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12503866?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12503866?buildTab=artifacts#/) on master @ [a0577c984accd44158a16db0f6c30f3fb349db2f](https://github.com/cockroachdb/cockroach/commits/a0577c984accd44158a16db0f6c30f3fb349db2f):


```
=== RUN   TestClusterResetSQLStats
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/03e3c3f86914ef5ebcca1a9ffbd26abe/logTestClusterResetSQLStats2329662074
    test_log_scope.go:81: use -show-logs to present logs inline
    stats_test.go:392: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/03e3c3f86914ef5ebcca1a9ffbd26abe/logTestClusterResetSQLStats2329662074
--- FAIL: TestClusterResetSQLStats (60.36s)
=== RUN   TestClusterResetSQLStats/flushed=true
    test_server_shim.go:165: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    stats_test.go:358: expected to find stats for at least one statement, but found: 0
    --- FAIL: TestClusterResetSQLStats/flushed=true (30.12s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #113828 pkg/server/application_api/application_api_test: TestClusterResetSQLStats failed [C-test-failure O-robot branch-release-23.2 release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterResetSQLStats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/server/application_api/application_api_test.TestClusterResetSQLStats [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12503867?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12503867?buildTab=artifacts#/) on master @ [a0577c984accd44158a16db0f6c30f3fb349db2f](https://github.com/cockroachdb/cockroach/commits/a0577c984accd44158a16db0f6c30f3fb349db2f):


```
=== RUN   TestClusterResetSQLStats
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/03e3c3f86914ef5ebcca1a9ffbd26abe/logTestClusterResetSQLStats4152677159
    test_log_scope.go:81: use -show-logs to present logs inline
    stats_test.go:392: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/03e3c3f86914ef5ebcca1a9ffbd26abe/logTestClusterResetSQLStats4152677159
--- FAIL: TestClusterResetSQLStats (52.49s)
=== RUN   TestClusterResetSQLStats/flushed=true
    test_server_shim.go:165: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    stats_test.go:358: expected to find stats for at least one statement, but found: 0
    --- FAIL: TestClusterResetSQLStats/flushed=true (25.06s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #113828 pkg/server/application_api/application_api_test: TestClusterResetSQLStats failed [C-test-failure O-robot branch-release-23.2 release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterResetSQLStats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/server/application_api/application_api_test.TestClusterResetSQLStats [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12516914?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12516914?buildTab=artifacts#/) on master @ [68d6832f09cb1be68a45dc930ee6366b45d16467](https://github.com/cockroachdb/cockroach/commits/68d6832f09cb1be68a45dc930ee6366b45d16467):


```
=== RUN   TestClusterResetSQLStats
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/03e3c3f86914ef5ebcca1a9ffbd26abe/logTestClusterResetSQLStats3119226648
    test_log_scope.go:81: use -show-logs to present logs inline
    stats_test.go:392: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/03e3c3f86914ef5ebcca1a9ffbd26abe/logTestClusterResetSQLStats3119226648
--- FAIL: TestClusterResetSQLStats (80.62s)
=== RUN   TestClusterResetSQLStats/flushed=true
    test_server_shim.go:165: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    stats_test.go:358: expected to find stats for at least one statement, but found: 0
    --- FAIL: TestClusterResetSQLStats/flushed=true (47.29s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #113828 pkg/server/application_api/application_api_test: TestClusterResetSQLStats failed [C-test-failure O-robot branch-release-23.2 release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterResetSQLStats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/backupccl/backup_test.go
+++ b/pkg/ccl/backupccl/backup_test.go
@@ -261,9 +261,6 @@ func TestBackupRestoreJobTagAndLabel(t *testing.T) {
 	tc, _, _, cleanupFn := backupRestoreTestSetupWithParams(t, numNodes, numAccounts, InitManualReplication,
 		base.TestClusterArgs{
 			ServerArgs: base.TestServerArgs{
-				DefaultTestTenant: base.TestDoesNotWorkWithSharedProcessModeButWeDontKnowWhyYet(
-					base.TestTenantProbabilistic, 113857, /* issueNumber */
-				),
 				Knobs: base.TestingKnobs{
 					DistSQL: &execinfra.TestingKnobs{
 						SetupFlowCb: func(ctx context.Context, _ base.SQLInstanceID, _ *execinfrapb.SetupFlowRequest) error {

--- a/pkg/server/application_api/stats_test.go
+++ b/pkg/server/application_api/stats_test.go
@@ -330,9 +330,6 @@ func TestClusterResetSQLStats(t *testing.T) {
 		t.Run(fmt.Sprintf("flushed=%t", flushed), func(t *testing.T) {
 			testCluster := serverutils.StartCluster(t, 3 /* numNodes */, base.TestClusterArgs{
 				ServerArgs: base.TestServerArgs{
-					DefaultTestTenant: base.TestDoesNotWorkWithSharedProcessModeButWeDontKnowWhyYet(
-						base.TestTenantProbabilistic, 113856, /* issueNumber */
-					),
 					Insecure: true,
 					Knobs: base.TestingKnobs{
 						SQLStatsKnobs: sqlstats.CreateTestingKnobs(),

--- a/pkg/server/testserver.go
+++ b/pkg/server/testserver.go
@@ -646,27 +646,29 @@ func (ts *testServer) startDefaultTestTenant(
 	return ts.StartTenant(ctx, params)
 }
 
-func (ts *testServer) startSharedProcessDefaultTestTenant(
-	ctx context.Context,
-) (serverutils.ApplicationLayerInterface, error) {
-	params := base.TestSharedProcessTenantArgs{
+func (ts *testServer) getSharedProcessDefaultTenantArgs() base.TestSharedProcessTenantArgs {
+	args := base.TestSharedProcessTenantArgs{
 		TenantName:  "test-tenant",
 		TenantID:    serverutils.TestTenantID(),
 		Knobs:       ts.params.Knobs,
 		UseDatabase: ts.params.UseDatabase,
 	}
 	// See comment above on separate process tenant regarding the testing knobs.
-	params.Knobs.Server = &TestingKnobs{}
+	args.Knobs.Server = &TestingKnobs{}
 	if ts.params.Knobs.Server != nil {
-		params.Knobs.Server.(*TestingKnobs).DiagnosticsTestingKnobs = ts.params.Knobs.Server.(*TestingKnobs).DiagnosticsTestingKnobs
+		args.Knobs.Server.(*TestingKnobs).DiagnosticsTestingKnobs = ts.params.Knobs.Server.(*TestingKnobs).DiagnosticsTestingKnobs
 	}
+	return args
+}
 
-	tenant, _, err := ts.StartSharedProcessTenant(ctx, params)
+func (ts *testServer) startSharedProcessDefaultTestTenant(
+	ctx context.Context,
+) (serverutils.ApplicationLayerInterface, error) {
+	tenant, _, err := ts.StartSharedProcessTenant(ctx, ts.getSharedProcessDefaultTenantArgs())
 	if err != nil {
 		return nil, err
 	}
-
-	return tenant, err
+	return tenant, nil
 }
 
 // maybeStartDefaultTestTenant might start a test tenant. This can then be used
@@ -790,6 +792,18 @@ func (ts *testServer) grantDefaultTenantCapabilities(
 // The caller is responsible for calling .Stopper().Stop() even
 // when PreStart() returns an error.
 func (ts *testServer) PreStart(ctx context.Context) error {
+	// In case we'll need to start the shared-process default test tenant later
+	// down the line, make sure that we set the correct arguments. This matters
+	// in multi-node clusters where we need this to happen before the first call
+	// to Activate in order to prevent the race between testServer.Activate
+	// explicitly starting the shared-process tenant with the correct args and
+	// the server controller realizing that it's missing a tenant and starting
+	// one with no test args.
+	func(args base.TestSharedProcessTenantArgs) {
+		ts.topLevelServer.serverController.mu.Lock()
+		defer ts.topLevelServer.serverController.mu.Unlock()
+		ts.topLevelServer.serverController.mu.testArgs[args.TenantName] = args
+	}(ts.getSharedProcessDefaultTenantArgs())
 	return ts.topLevelServer.PreStart(ctx)
 }
 

--- a/pkg/sql/execstats/traceanalyzer_test.go
+++ b/pkg/sql/execstats/traceanalyzer_test.go
@@ -57,9 +57,6 @@ func TestTraceAnalyzer(t *testing.T) {
 	tc := serverutils.StartCluster(t, numNodes, base.TestClusterArgs{
 		ReplicationMode: base.ReplicationManual,
 		ServerArgs: base.TestServerArgs{
-			DefaultTestTenant: base.TestDoesNotWorkWithSharedProcessModeButWeDontKnowWhyYet(
-				base.TestTenantProbabilistic, 113598,
-			),
 			UseDatabase: "test",
 			Knobs: base.TestingKnobs{
 				SQLExecutor: &sql.ExecutorTestingKnobs{

--- a/pkg/sql/sqlstats/persistedsqlstats/datadriven_test.go
+++ b/pkg/sql/sqlstats/persistedsqlstats/datadriven_test.go
@@ -75,9 +75,6 @@ func TestSQLStatsDataDriven(t *testing.T) {
 
 	ctx := context.Background()
 	var params base.TestServerArgs
-	params.DefaultTestTenant = base.TestDoesNotWorkWithSharedProcessModeButWeDontKnowWhyYet(
-		base.TestTenantProbabilistic, 113854, /* issueNumber */
-	)
 	knobs := sqlstats.CreateTestingKnobs()
 	knobs.StubTimeNow = stubTime.Now
 	knobs.OnStmtStatsFlushFinished = injector.invokePostStmtStatsFlushCallback

--- a/pkg/sql/sqlstats/persistedsqlstats/flush_test.go
+++ b/pkg/sql/sqlstats/persistedsqlstats/flush_test.go
@@ -74,9 +74,6 @@ func TestSQLStatsFlush(t *testing.T) {
 
 	testCluster := serverutils.StartCluster(t, 3 /* numNodes */, base.TestClusterArgs{
 		ServerArgs: base.TestServerArgs{
-			DefaultTestTenant: base.TestDoesNotWorkWithSharedProcessModeButWeDontKnowWhyYet(
-				base.TestTenantProbabilistic, 113855, /* issueNumber */
-			),
 			Knobs: base.TestingKnobs{
 				SQLStatsKnobs: &sqlstats.TestingKnobs{
 					StubTimeNow: fakeTime.Now,

